libraries/AP_GPS/AP_GPS_SBF.cpp

```diff
@@ -0,0 +1,368 @@
+        msg4007* temp = (msg4007*)sbf_msg.data;
+
+        // Update time state
+        if (temp->WNc != 65535) {
+            state.time_week = temp->WNc;
+            state.time_week_ms = (uint32_t)(temp->TOW);
+        }
+		
+		state.last_gps_time_ms = hal.scheduler->millis();
+
+        state.hdop = last_hdop;
+
+        // Update velocity state (dont use −2·10^10)
+        if (temp->Vn > -20000000000) {
+            state.velocity.x = (float)(temp->Vn);
+            state.velocity.y = (float)(temp->Ve);
+            state.velocity.z = (float)(-temp->Vu);
+			
+			state.have_vertical_velocity = true;
+
+            float ground_vector_sq = state.velocity[0] * state.velocity[0] + state.velocity[1] * state.velocity[1];
+            state.ground_speed = (float)safe_sqrt(ground_vector_sq);
+
+            state.ground_course_cd = (int32_t)(100 * ToDeg(atan2f(state.velocity[1], state.velocity[0])));
+            if (state.ground_course_cd < 0) {
+                state.ground_course_cd += 36000;
+            }
+        }
+
+        // Update position state (dont use −2·10^10)
+        if (temp->Latitude > -20000000000) {
+            state.location.lat = (int32_t)(temp->Latitude * 57.295779513 * 1e7);
+            state.location.lng = (int32_t)(temp->Longitude * 57.295779513 * 1e7);
+            state.location.alt = (int32_t)(temp->Height * 1e2f);
+        }
+
+        if (temp->NrSV != 255) {
+            state.num_sats = temp->NrSV;
+        }
+
+        Debug("temp->Mode=0x%02x\n", (unsigned)temp->Mode);
+        switch (temp->Mode & 15) {
+            case 0: // no pvt
+                state.status = AP_GPS::NO_FIX;
+                break;
+            case 1: // standalone
+                state.status = AP_GPS::GPS_OK_FIX_3D;
+                break;
+            case 2: // dgps
+                state.status = AP_GPS::GPS_OK_FIX_3D_DGPS;
+                break;
+            case 3: // fixed location
+                state.status = AP_GPS::GPS_OK_FIX_3D;
+                break;
+            case 4: // rtk fixed
+                state.status = AP_GPS::GPS_OK_FIX_3D_RTK;
+                break;
+            case 5: // rtk float
+                state.status = AP_GPS::GPS_OK_FIX_3D_DGPS;
+                break;
+            case 6: // sbas
+                state.status = AP_GPS::GPS_OK_FIX_3D;
+                break;
+            case 7: // moving rtk fixed
+                state.status = AP_GPS::GPS_OK_FIX_3D_RTK;
+                break;
+            case 8: // moving rtk float
+                state.status = AP_GPS::GPS_OK_FIX_3D_DGPS;
+                break;
+        }
+        
+        if ((temp->Mode & 64) > 0) // gps is in base mode
+            state.status = AP_GPS::NO_FIX;
+        if ((temp->Mode & 128) > 0) // gps only has 2d fix
+            state.status = AP_GPS::GPS_OK_FIX_2D;
+                    
+        return true;
+    }
+    // DOP
+    if (blockid == 4001) {
+        msg4001* temp = (msg4001*)sbf_msg.data;

```

Comment:
why not use a union like we use in other GPS drivers instead of having these casts?